### PR TITLE
Thread-per-node stencil operator kernels

### PR DESCRIPTION
**Pull request by sriharshakandala**
- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: src/Operators/finitedifference.jl [hunk 1]
```diff
@@ -3437,26 +3436,43 @@ function Base.copyto!(
         Nq = 1
         Nh = 1
     end
-    bounds = window_bounds(space, bc)
-    # executed
-    @cuda threads = (Nq, Nq) blocks = (Nh,) copyto_stencil_kernel!(
+    (li, lw, rw, ri) = bounds = window_bounds(space, bc)
+    nnodes = ri - li + 1
+    max_threads = 256
+    nitems = nnodes * Nq * Nq * Nh # # of independent interior items
```
Comment (bloops) [inline src/Operators/finitedifference.jl:3442]: nit: `# number of independent interior items`

**Review by bloops — APPROVED** [review]

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: src/Operators/finitedifference.jl [hunk 2]
```diff
@@ -3437,26 +3436,43 @@ function Base.copyto!(
...
-    h = blockIdx().x
-    hidx = (i, j, h)
-    apply_stencil!(space, out, bc, hidx, bds)
+function copyto_stencil_kernel!(out, bc, space, bds, Nq, Nh, Nv)
+    gid = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+    if gid ≤ Nv * Nq * Nq * Nh
+        (li, lw, rw, ri) = bds
+        (v, i, j, h) = Spaces._get_idx((Nv, Nq, Nq, Nh), gid)
+        hidx = (i, j, h)
+        idx = (li:ri)[v]
```
Comment (simonbyrne) [inline src/Operators/finitedifference.jl:3462]: actually i think it might be better to do
```suggestion
        idx = v - 1 + li
```

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**